sdcflows/fieldmaps.py

```diff
@@ -0,0 +1,345 @@
+
+    bids_root = attr.ib(init=False, default=None, repr=False)
+    """Path of the BIDS root."""
+
+    metadata = attr.ib(kw_only=True, default=attr.Factory(dict))
+    """
+    Metadata associated to this file. When provided as keyword argument in initialization,
+    will overwrite metadata read from sidecar JSON.
+    """
+
+    find_meta = attr.ib(kw_only=True, default=True, type=bool, repr=False)
+    """Enable/disable automated search for corresponding metadata."""
+
+    @path.validator
+    def check_path(self, attribute, value):
+        """Validate a fieldmap path."""
+        if isinstance(value, BIDSFile):
+            value = Path(value.path)
+        if isinstance(value, str):
+            value = Path(value)
+
+        if not value.is_file():
+            raise FileNotFoundError(
+                f"File path <{value}> does not exist, is a broken link, or it is not a file"
+            )
+
+        if not str(value).endswith((".nii", ".nii.gz")):
+            raise ValueError(f"File path <{value}> does not look like a NIfTI file.")
+
+        suffix = str(value).rpartition(".nii")[0].rpartition("_")[-1]
+        if suffix not in tuple(MODALITIES.keys()):
+            raise ValueError(
+                f"File path <{value}> with suffix <{suffix}> is not a valid "
+                "fieldmap sourcefile."
+            )
+
+    def __attrs_post_init__(self):
+        """Validate metadata and additional checks."""
+        self.entities = parse_file_entities(str(self.path))
+        self.suffix = self.entities.pop("suffix")
+        extension = self.entities.pop("extension").lstrip(".")
+
+        # Automatically fill metadata in when possible
+        # TODO: implement BIDS hierarchy of metadata
+        if self.find_meta:
+            sidecar = Path(str(self.path).replace(extension, "json"))
+            if sidecar.is_file():
+                _meta = self.metadata or {}
+                self.metadata = loads(sidecar.read_text())
+                self.metadata.update(_meta)
+
+        # Attempt to infer a bids_root folder
+        relative_path = relative_to_root(self.path)
+        if str(relative_path) != str(self.path):
+            self.bids_root = Path(str(self.path)[: -len(str(relative_path))])
+
+        # Check for REQUIRED metadata (depends on suffix.)
+        if self.suffix in ("bold", "dwi", "epi", "sbref"):
+            if "PhaseEncodingDirection" not in self.metadata:
+                raise MetadataError(
+                    f"Missing 'PhaseEncodingDirection' for <{self.path}>."
+                )
+            if not (
+                set(("TotalReadoutTime", "EffectiveEchoSpacing")).intersection(
+                    self.metadata.keys()
+                )
+            ):
+                raise MetadataError(
+                    f"Missing readout timing information for <{self.path}>."
+                )
+
+        if self.suffix == "fieldmap" and "Units" not in self.metadata:
```

Comment:
no need to hit anymore lines
```suggestion
        elif self.suffix == "fieldmap" and "Units" not in self.metadata:
```